Improve documentation and error messaging for local extension installations without executables

BagToad opened the pull request:
Fixes #9760 

## Requirements

> ### Acceptance Criteria
> 
> **Given** I am in a directory that doesn't have an executable for `gh` to call as an extension
> **When** I run `gh extension install .`
> **Then** I get an informative message on stderr saying "An extension has been installed but there is no executable found in this directory, perhaps you need to build it?"
> 
> This should be styled when there is a TTY and unstyled when there is not.
> 
> **When** I run `gh extension install --help`
> **Then** I see an explanation that `gh extension install .` creates a symlink to an executable in the local directory regardless of whether it exists or not, and it's necessary for the user to build executables out of band if need be.

## Demo

![image](https://github.com/user-attachments/assets/3c7e411a-d331-4959-9850-22e4dea78d6d)

## Reproduction

**Error message:**

```shell
# Setup path to `gh` binary built from this PR
bingh="<PATH_TO_GH_BINARY>"

$bingh ext create testing-9933
cd gh-testing-9933
rm gh-testing-9933
$bingh ext install .
```

**Cleanup/reset:**

```shell
# Assuming you're still in the testing extension dir...
cd ..
$bingh ext remove gh-testing-9933
rm -rf gh-testing-9933
```

BagToad committed:
Warn when installing local ext with no executable

BagToad committed:
Document requirements for local extensions

jtmcg commented on pkg/cmd/extension/command.go:
```diff
@@ -293,19 +293,35 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
 				Short: "Install a gh extension from a repository",
 				Long: heredoc.Docf(`
-					Install a GitHub repository locally as a GitHub CLI extension.
+					Install a GitHub CLI extension from a GitHub repository.
```
I appreciate the cleanup of the language, here. I have a suggestion on how to make this even better

```suggestion
					Install a GitHub CLI extension from a GitHub or local repository.
```

The value prop here is that there are two ways to install an extension, which we call out here. In the comments further down, we specify which of the two ways we are providing instructions for and useful information about them.

What do you think?

BagToad replied:
I like this improvement a lot, thanks @jtmcg ✨

jtmcg commented on pkg/cmd/extension/command.go:
```diff
@@ -293,19 +293,35 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
+					Install a GitHub CLI extension from a GitHub repository.
 
-					The repository argument can be specified in %[1]sOWNER/REPO%[1]s format as well as a full URL.
+					The repository argument can be specified in %[1]sOWNER/REPO%[1]s format or as a full repository URL.
```
Continued from above
```suggestion
					For GitHub repositories, the repository argument can be specified in %[1]sOWNER/REPO%[1]s format or as a full repository URL.
```

jtmcg commented on pkg/cmd/extension/command.go:
```diff
@@ -293,19 +293,35 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
-					To install an extension in development from the current directory, use %[1]s.%[1]s as the
-					value of the repository argument.
+					To install an in-development extension from a locally cloned repository, use %[1]s.%[1]s as the
+					value of the repository argument. Note the following:
```
Continued from above
```suggestion
					For local repositories, often used while developing extensions, use %[1]s.%[1]s as the
					value of the repository argument. Note the following:
```

jtmcg commented on pkg/cmd/extension/manager.go:
```diff
@@ -194,10 +195,24 @@ func (m *Manager) populateLatestVersions(exts []*Extension) {
+	// it does indicate that the user will not be able to run the extension until
+	// the executable file is built or created manually somehow.
+	if _, err := os.Stat(filepath.Join(dir, name)); err != nil {
+		errMsg := fmt.Errorf("%v %w: expected executable file named \"%s\" in %s, perhaps you need to build it?", cs.WarningIcon(), ErrExtensionExecutableNotFound, name, dir)
```
🤔 After reading the comment above, I'm wondering if adding something informing the user that the extension won't run would be helpful: 
```suggestion
		errMsg := fmt.Errorf("%v %w: executable file named \"%s\" in %s is required for the extension to run after install. Perhaps you need to build it?", cs.WarningIcon(), ErrExtensionExecutableNotFound, name, dir)
```

BagToad replied:
Sure, I don't see any problems with this 👍

jtmcg commented on pkg/cmd/extension/command.go:
```diff
@@ -293,19 +293,35 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
+					- After installing an extension from a locally cloned repository, the GitHub CLI will
+					manage this extension as a symbolic link pointing to an executable file with the same
+					name as the repository in the repository's root.
+						- For example, if the repository is named %[1]sgh-foobar%[1]s, the symbolic link will
```
From the tests below, it seems like what's described here is the non-windows behavior:

> // We don't create symlinks on Windows, so check if we made a file
		// with the correct contents

How is windows different and should we enumerate those differences here?

BagToad replied:
Really good question 🤔 

I think how we handle Windows symlinks could be seen as an implementation detail, and we _might_ be fine to just leave it out of the docs. 

On Windows, we create a file that contains a string indicating the path to the extension. So, it "behaves" like a symlink. I'm using "behaves" loosely, but what we really should say is that we _use_ this file as if it were a symlink. Given the "behavior" is the same, I'm okay not mentioning it, but I could also see value in documenting this. With all that said, what do you think we should do?

Here is[ that source code for windows "symlinks"](https://github.com/cli/cli/blob/9decf1b5260771ae882fa6f8191794d918b46465/pkg/cmd/extension/symlink_windows.go) for more context. 

The [test I wrote](https://github.com/cli/cli/blob/c5497b4d28f98ea9d626ef61190dbdfe8e67b89d/pkg/cmd/extension/manager_test.go#L739-L758) needs to accommodate this Windows behavior too, so that might also help communicate the Windows'isms.

jtmcg replied:
I find the idea that the test works as documentation of it fairly compelling. Since it's symlink-like for windows, there's probably no need to enumerate the semantics in our docs. Notably, I also didn't pick up on the windows difference until I saw the test, so relying on the test for that documentation seems reasonable, since the answer to what we do is right there in the test. Maybe we update the test comment to be a bit more specific? I've taken a stab at it: https://github.com/cli/cli/pull/9933/files#r1847314568

BagToad replied:
Sounds good to me! Thanks for the suggestion ✨ , I committed it.

jtmcg commented on pkg/cmd/extension/manager_test.go:
```diff
@@ -719,6 +719,62 @@ func TestManager_UpgradeExtension_GitExtension_Pinned(t *testing.T) {
 	gcOne.AssertExpectations(t)
 }
 
+func TestManager_Install_local(t *testing.T) {
```
❤️

jtmcg reviewed (commented):
I've tested and it all looks good. I appreciate the updates to the docs language 🙌 I've added some thoughts on how to possibly improve them further.

jtmcg commented on pkg/cmd/extension/manager_test.go:
```diff
@@ -719,6 +719,62 @@ func TestManager_UpgradeExtension_GitExtension_Pinned(t *testing.T) {
+	extensionLinkFile := filepath.Join(extManagerDir, "extensions", fakeExtensionName)
+
+	if runtime.GOOS == "windows" {
+		// We don't create symlinks on Windows, so check if we made a file
+		// with the correct contents
```
Follow up from [docs comment](https://github.com/cli/cli/pull/9933/files#r1847313344) above

```suggestion
		// We don't create true symlinks on Windows, so check if we made a
		// file with the correct contents to produce the symlink-like behavior
```

BagToad committed:
Update test comments about Windows behavior

Co-authored-by: Tyler McGoffin <jtmcg@github.com>

BagToad committed:
Update language for missing extension executable

Co-authored-by: Tyler McGoffin <jtmcg@github.com>

BagToad committed:
Improve docs on installing extensions

Better clarify the two extension types in the `extension install` docs.

Co-authored-by: Tyler McGoffin <jtmcg@github.com>

williammartin commented:
It occurs to me that this is a breaking change. I wonder if there are scripts out there that do:

```
gh ext install .
go build ...
```

Which will start breaking. Need to think about this.

williammartin commented:
So we have two options here:
 1. We just ship this and if someone complains we either say "sorry" and revert it, or say "sorry" and they hopefully make an easy change to their script
 2. We only show this message in interactive mode, which is probably the most common

The advantage of not breaking is clear, the advantage of breaking is consistency. I'd probably favour consistency and if it breaks someone we can apologise. Happy for someone to disagree though.

BagToad commented:
> It occurs to me that this is a breaking change. I wonder if there are scripts out there that do:
> 
> ```
> gh ext install .
> go build ...
> ```
> 
> Which will start breaking. Need to think about this.

@williammartin could you talk more about why this would be breaking? Apologies if it is obvious 😅 

Are you talking about if someone had a script checking if `gh ext install .` is successful, then it's no longer going to be successful because we're sending a message out on stderr?

Maybe we only send that message to stderr when there is a TTY?

williammartin commented:
@BagToad my bad, I actually thought this was erroring (exit non-zero), not just a warning. I don't feel strongly either way about just the message. Maybe I'd favour just having it on TTY, since realistically, people who are running it in a script probably understand what the semantics are of the command.

BagToad commented:
@williammartin, I think you're right actually. We are returning an `error` now and that makes us exit non-zero:

```
! an extension has been installed but there is no executable: expected executable file named "gh-testing-9933" in /Users/kynanware/Documents/orgs/cli/gh-testing-9933, perhaps you need to build it?

❯ echo $?
1
```

Ideas:

- Only return the error on TTY, so we only exit non-zero with a TTY. 
    - I think this one is a bit confusing - I don't like having a different exit code whether there is a TTY or not.  
- We do not print this message to stderr or return an err like originally written in the A/C. Instead, we print it to stdout when there is a TTY. Exit code remains zero for this case.

williammartin commented:
I think we should continue to exit 0 in all cases. I think most uses of `cs.WarningIcon()` (which is the icon for `!`) go to ErrOut. I'd probably be inclined to continue sending it to that stream.

BagToad committed:
Change: exit zero, still print warning to stderr

Instead of returning `ErrExtensionExecutableNotFound` error which causes `gh` to have a non-zero exit code, catch it and print the message to stderr, returning nil. Only print the warning to stderr when there is a TTY.

BagToad committed:
Update error message wording

BagToad committed:
Assert stderr is empty in manager_test.go

BagToad commented:
Thanks @williammartin 🙇 

I pushed a few commits to fix that if you could give it another review, please. 

I also added tests in `command_test.go` to demonstrate this additional behavior.

## Demo

No TTY
```
❯ $bingh ext install . | cat    

❯ echo $?
0
```

With TTY

```
❯ $bingh ext install .          
! an extension has been installed but there is no executable: executable file named "gh-testing-9933" in /Users/kynanware/Documents/orgs/cli/gh-testing-9933 is required to run the extension after install. Perhaps you need to build it?

❯ echo $?
0
```

### Comments:

I saw inconsistency in the way the rest of the codebase checks if we are in a TTY before printing to stderr. I saw three different approaches:

**1. Check if stderr is TTY only, then print to stderr:**

https://github.com/cli/cli/blob/9177b22a213318ccc00a9b08fbac1dd14fb3a839/pkg/cmd/gist/view/view.go#L108-L110

**2. Check if stdout is TTY only, then print to stderr:**

https://github.com/cli/cli/blob/9177b22a213318ccc00a9b08fbac1dd14fb3a839/pkg/cmd/extension/command.go#L168-L170

**3. Check if both stdout and stderr are TTY, then print to stderr:**

https://github.com/cli/cli/blob/9177b22a213318ccc00a9b08fbac1dd14fb3a839/pkg/cmd/repo/fork/fork.go#L187
https://github.com/cli/cli/blob/9177b22a213318ccc00a9b08fbac1dd14fb3a839/pkg/cmd/repo/fork/fork.go#L213-L218

---

With all the different options, I chose option _2 - check if stdout is TTY only, then print to stderr_ because there is precedent for this already in `extension/command.go` (see example linked in point 2 above 👆 ). Snippet of what I proposed:

https://github.com/cli/cli/blob/e5eedefd093afa784d5f39323be31916738d02a3/pkg/cmd/extension/command.go#L344-L346

Let me know if you think I should have chosen a different option, please.

williammartin commented on pkg/cmd/extension/command.go:
```diff
@@ -293,19 +293,35 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
 				Short: "Install a gh extension from a repository",
 				Long: heredoc.Docf(`
-					Install a GitHub repository locally as a GitHub CLI extension.
+					Install a GitHub CLI extension from a GitHub or local repository.
```
I don't think there's anything that requires the local directory to be a repository, so this doesn't seem quite right.

That said I understand that the short and use indicate that it's a repository, so let's just leave it alone.

BagToad replied:
I had this thought too and here is how I rationalized it to myself:

I saw us using `repository` to describe the directory as a suggestion that we want extension authors to consider their extensions as repositories, even if we don't strictly require it. We especially want them to consider them as _GitHub_ repositories, except in this case it is cloned locally.

Maybe there's holes on my logic 🤪 , but that was my thoughts and why I didn't change it.

I'll mark this thread as resolved, but feel free to re-open it if you think we have something to do here 🙏

williammartin commented on pkg/cmd/extension/command.go:
```diff
@@ -293,19 +293,35 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
+					For local repositories, often used while developing extensions, use %[1]s.%[1]s as the
+					value of the repository argument. Note the following:
+
+					- After installing an extension from a locally cloned repository, the GitHub CLI will
```
Do you think it's worth doc-ing how this is different on windows?

BagToad replied:
@jtmcg and I [also wondered about that](https://github.com/cli/cli/pull/9933#discussion_r1847049096) but landed on it being an implementation detail that may confuse the docs if we try to document it. 

What do you think?

BagToad replied:
@williammartin and I talked about this and decided to add a hint to the help text without fully documenting the differences.

BagToad replied:
Resolved in 32abca8c381c34a349d7d829d2fcfda2a4da91e6

williammartin commented on pkg/cmd/extension/manager.go:
```diff
@@ -194,10 +195,24 @@ func (m *Manager) populateLatestVersions(exts []*Extension) {
+	// An error here doesn't indicate a failed extension installation, but
+	// it does indicate that the user will not be able to run the extension until
+	// the executable file is built or created manually somehow.
+	if _, err := os.Stat(filepath.Join(dir, name)); err != nil {
```
This isn't the only reason that `err` might occur, so if we want to be specific, we should check [`os.IsNotExist`](https://pkg.go.dev/os#IsNotExist).

BagToad replied:
Done in b5f3463b7530f4e1286a5bb3b5e0c4c4eb22309b

williammartin commented on pkg/cmd/extension/manager.go:
```diff
@@ -194,10 +195,24 @@ func (m *Manager) populateLatestVersions(exts []*Extension) {
+	// it does indicate that the user will not be able to run the extension until
+	// the executable file is built or created manually somehow.
+	if _, err := os.Stat(filepath.Join(dir, name)); err != nil {
+		errMsg := fmt.Errorf("%v %w: executable file named \"%s\" in %s is required to run the extension after install. Perhaps you need to build it?", cs.WarningIcon(), ErrExtensionExecutableNotFound, name, dir)
```
Maybe just return `fmt.Errorf`, since the variable is only used once, and it's not really a message, it's an error.

You could also avoid this whole wrapping business and do:

```go
type ErrExtensionExecutableNotFound struct {
 Dir string
 Name string
}
```

And either construct the error in `Error` method, or at the call site, if that's the only place it is used.

No particularly strong feelings here though, just offering an alternative idea.

BagToad replied:
Thanks @williammartin 🙇 

I took a stab at your suggested alternative in b2ab7b84f84887e2757fe28be6ca616e2582f853 - I think it looks better in general

williammartin reviewed (commented):
Generally looking really good.

BagToad committed:
Improve error handling for missing executable

Check for executable file existence using os.IsNotExist for clearer error handling

jtmcg commented on pkg/cmd/extension/command.go:
```diff
@@ -322,7 +338,15 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
-						return m.InstallLocal(wd)
+
+						err = m.InstallLocal(wd)
+						if errors.Is(err, ErrExtensionExecutableNotFound) {
+							if io.IsStdoutTTY() {
+								fmt.Fprintln(io.ErrOut, err.Error())
+							}
+							return nil
```
Are there any circumstances where it would be helpful to print this error if not TTY? I imagine someone installing an extension as part of a workflow, this error occurring, and them lacking any useful information to debug it. I did a [code search for `gh extension install`](https://github.com/search?q=%22gh+extension+install%22&type=code) and see a couple examples of this on the first results page.
```suggestion
						if errors.Is(err, ErrExtensionExecutableNotFound) {
							fmt.Fprintln(io.ErrOut, err.Error())
						        return nil
						}
```

BagToad replied:
@williammartin [mentioned that](https://github.com/cli/cli/pull/9933#issuecomment-2486418124), and this was his reasoning to prefer having the message only on TTY:

>  I don't feel strongly either way about just the message. Maybe I'd favour just having it on TTY, since realistically, people who are running it in a script probably understand what the semantics are of the command.

---

> I did a [code search for gh extension install](https://github.com/search?q=%22gh+extension+install%22&type=code) and see a couple examples of this on the first results page

I think a better code search is to [search for `gh extension install .`](https://github.com/search?q=%22gh+extension+install+.%22&type=code) instead, since this message can only happen with this syntax to install a local extension.

We could [take the search a bit further as well](https://github.com/search?q=%22gh+extension+install+.%22+-path%3A*.md+AND+-path%3A*.go&type=code) by removing markdown and Go files, focusing on identifying automations where an extension might be installed locally. I removed markdown from the search because we can loosely assume markdown instructions are for user guided programs and would have a TTY. I removed Go files because it doesn't make sense for this to be in Go files aside from forks of `cli/cli`.

And if you want to see actions workflows specifically, look at [this search](https://github.com/search?q=%22gh+extension+install+.%22+-path%3A*.md+AND+-path%3A*.go+AND+path%3A**.github%2Fworkflows%2F*&type=code).

Either way, I think it is worth saying that this error is not actually an error - it's actually a warning. It does not indicate a failure to install the extension; it is a warning that additional steps are required to have the installed extension work. These additional steps already exist, so existing automations and workflows are likely unaffected and don't need to see this new warning message. I suppose the argument is that someone authoring a new automation or workflow might be interested in seeing this new warning message, or than an existing workflow breaks and it might be useful for them to see this new warning message 🤷 

Though, the user/script definitely does receive an error message _later_ when they attempt to run an extension that has no executable 😜 

I suppose the last consideration I'll throw into the mix is that we have some level of precedent in `extension.go` for showing informative messages to the user only when there is a TTY (though I can see the argument that our case is different):

https://github.com/cli/cli/blob/9177b22a213318ccc00a9b08fbac1dd14fb3a839/pkg/cmd/extension/command.go#L168-L170

I don't care particularly strongly. Thoughts?

jtmcg replied:
Makes sense. That search you've put together is much better than mine and really shows the space I have to level-up using GH code search 😅 Thanks for taking the time to show me this 🙌 

> I suppose the argument is that someone authoring a new automation or workflow might be interested in seeing this new warning message, or than an existing workflow breaks and it might be useful for them to see this new warning message 🤷

I think this is probably the most useful use-case, but that a new author is likely going to be testing installs TTY and not through automation, so they'll get the error message anyway. Let's leave it as it is, then. If we get feedback from folks that they could have benefited from more guidance, we can revisit 👍

BagToad committed:
Refactor extension executable error handling

williammartin commented on pkg/cmd/extension/command_test.go:
```diff
@@ -286,6 +286,49 @@ func TestNewCmdExtension(t *testing.T) {
+				return nil
+			},
+			wantStderr: fmt.Sprintf(
+				"! %s",
```
I would like for this test to be checking the hardcoded string directly, rather than relying on the implementation detail of `Error()`. This way, we have confidence that if the returned value from `Error()` was changed, something would tell us.

BagToad replied:
Resolved by ec4b2dfe1b6770845874bd78433c843083e0e2b3

williammartin commented on pkg/cmd/extension/manager.go:
```diff
@@ -194,10 +203,25 @@ func (m *Manager) populateLatestVersions(exts []*Extension) {
+	// An error here doesn't indicate a failed extension installation, but
+	// it does indicate that the user will not be able to run the extension until
+	// the executable file is built or created manually somehow.
+	if _, err := os.Stat(filepath.Join(dir, name)); os.IsNotExist(err) {
```
This will swallow the error if it's anything other than os.IsNotExist

BagToad replied:
Resolved by 3b93e28910a8902d348a24842e8f07f6738b88ba

BagToad committed:
Clarify hosts in ext install help text

BagToad committed:
Assert on err msg directly in ext install tests

BagToad committed:
Fix error mishandling in local ext install

BagToad committed:
Mention Windows quirk in ext install help text

williammartin committed:
Reformat ext install long

williammartin commented:
After some discussion we decided that the message should only be displayed when there is a TTY.

 ### Acceptance Criteria

**Given** I invoke `gh` without a TTY
**And Given** I am in a directory that doesn't have an executable for `gh` to call as an extension
**When** I run `gh extension install .`
**Then** I get no output (as is current behaviour)

```
~/workspace/cli/bin/gh ext install . | cat
```

**Given** I invoke `gh` with a TTY
**And Given** I am in a directory that doesn't have an executable for `gh` to call as an extension
**When** I run `gh extension install .`
**Then** I get an informative message on stderr saying "An extension has been installed but there is no executable found in this directory, perhaps you need to build it?"

```
➜ ~/workspace/cli/bin/gh ext install .
! an extension has been installed but there is no executable: executable file named "gh-local" in /Users/williammartin/workspace/cli-triaging/local-ext/gh-local is required to run the extension after install. Perhaps you need to build it?
```

This should be styled when there is a TTY and unstyled when there is not.

**When** I run `gh extension install --help`
 **Then** I see an explanation that `gh extension install .` creates a symlink to an executable in the local directory regardless of whether it exists or not, and it's necessary for the user to build executables out of band if need be.

```
➜  gh-local ~/workspace/cli/bin/gh  extension install --help
Install a GitHub CLI extension from a GitHub or local repository.

For GitHub repositories, the repository argument can be specified in `OWNER/REPO` format or as a full repository URL.
The URL format is useful when the repository is not hosted on github.com.

For local repositories, often used while developing extensions, use `.` as the
value of the repository argument. Note the following:

- After installing an extension from a locally cloned repository, the GitHub CLI will
manage this extension as a symbolic link (or equivalent mechanism on Windows) pointing
to an executable file with the same name as the repository in the repository's root.
For example, if the repository is named `gh-foobar`, the symbolic link will point
to `gh-foobar` in the extension repository's root.
- When executing the extension, the GitHub CLI will run the executable file found
by following the symbolic link. If no executable file is found, the extension
will fail to execute.
- If the extension is precompiled, the executable file must be built manually and placed
in the repository's root.

For the list of available extensions, see <https://github.com/topics/gh-extension>.
```

LGTM
